PWGJE/EMCALJetTasks/UserTasks/AliAnalysisTaskSubJetFraction.cxx

```diff
@@ -1833,76 +1788,136 @@ void AliAnalysisTaskSubJetFraction::Terminate(Option_t *)
 }
 
 ///////////Function to modify jets based on Toy Models/////////
-AliEmcalJet* AliAnalysisTaskSubJetFraction::ModifyJet(AliEmcalJet* Jet, Int_t JetContNb, TString Modification){
-
+fastjet::PseudoJet AliAnalysisTaskSubJetFraction::ModifyJet(AliEmcalJet* Jet, Int_t JetContNb, TString Modification){
   std::vector<fastjet::PseudoJet> fInputVectors;
   AliJetContainer *fJetCont = GetJetContainer(JetContNb);
   AliParticleContainer *fTrackCont = fJetCont->GetParticleContainer();
+  std::vector<fastjet::PseudoJet> Modified_Jet;
+  fastjet::ClusterSequence *fClustSeqSA;
+  Int_t NJetTracksTest = Jet->GetNumberOfTracks();
   if (fTrackCont) for (Int_t i=0; i<Jet->GetNumberOfTracks(); i++) {
       AliVParticle *fTrk = Jet->TrackAt(i, fTrackCont->GetArray());
       if (!fTrk) continue;    
       fastjet::PseudoJet PseudoTracks(fTrk->Px(), fTrk->Py(), fTrk->Pz(),fTrk->E());
       PseudoTracks.set_user_index(Jet->TrackAt(i)+100); //100 is very arbitary....why is it here anyway?
       fInputVectors.push_back(PseudoTracks);
     }
-  //  if (Modification=="Randomise") fInputVectors=RandomiseTracks(fInputVectors);
-  // if (Modification=="AddExtraProng_02_30") fInputVectors=AddExtraProng(fInputVectors,0.2,0.3);
-  AliEmcalJet *Modified_AliEmcalJet=0x0;
+  if (Modification=="Randomise") fInputVectors=RandomiseTracks(Jet,fInputVectors);
+  if (Modification=="AddExtraProng_02_15") fInputVectors=AddExtraProng(fInputVectors,0.2,0.15);
+  if (Modification=="AddExtraProng_02_30") fInputVectors=AddExtraProng(fInputVectors,0.2,0.30);
+  if (Modification=="AddExtraProng_02_45") fInputVectors=AddExtraProng(fInputVectors,0.2,0.45);
+  if (Modification=="AddExtraProng_02_60") fInputVectors=AddExtraProng(fInputVectors,0.2,0.60);
+  if (Modification=="AddExtraProng_01_30") fInputVectors=AddExtraProng(fInputVectors,0.1,0.30);
+  if (Modification=="AddExtraProng_01_45") fInputVectors=AddExtraProng(fInputVectors,0.1,0.45);
+  if (Modification=="AddExtraProng_03_30") fInputVectors=AddExtraProng(fInputVectors,0.3,0.30);
+  if (Modification=="AddExtraProng_03_45") fInputVectors=AddExtraProng(fInputVectors,0.3,0.45);
+  fJetRadius=fJetRadius*2.0;
   try {
-    fastjet::JetDefinition fJetDef(fastjet::antikt_algorithm, fJetRadius*2, static_cast<fastjet::RecombinationScheme>(0), fastjet::BestFJ30 );         
-    fastjet::ClusterSequence fClustSeqSA(fInputVectors, fJetDef);
-    std::vector<fastjet::PseudoJet> Modified_Jet=fClustSeqSA.inclusive_jets(0);
-    Modified_AliEmcalJet= new AliEmcalJet(Modified_Jet[0].perp(), Modified_Jet[0].pseudorapidity(), Modified_Jet[0].phi(), Modified_Jet[0].m());
+    fastjet::JetDefinition fJetDef(fastjet::antikt_algorithm, fJetRadius, static_cast<fastjet::RecombinationScheme>(0), fastjet::BestFJ30 );         
+    fClustSeqSA=new fastjet::ClusterSequence(fInputVectors, fJetDef);
```

Comment:
Why is this cluster sequence allocated on the heap?

Reply:
Hi Markus,

The reason is that to access a pseudojet's constituents the clustersequence needs to remain in scope (I was getting a fastjet error before). Since I need to access these constituents in functions outside this function I decided to allocate it on the heap so that it would remain in scope. 

Thanks,
Nima

Reply:
Hi @nzardosh ,

I see the problem, however this will lead again to a memory leak. In this case you should at least return something like a std::pair with the jet and its associated cluster sequence so you can delete the cluster sequence once you don't need the jet any more.

Cheers

Markus

Reply:
Hi Markus,

Ok I looked up std::pair last night and have sent a patch to update the code. Please check if this resolves the problem.

Thanks,
Nima

Reply:
Hi @nzardosh ,

Looks good (waiting for the result of the test, of course)! No further objections from my side!

Cheers

Markus